syncer/streamer_controller.go

```diff
@@ -0,0 +1,339 @@
+	initBinlogType BinlogType
+
+	// the current binlog type
+	currentBinlogType BinlogType
+
+	syncCfg        replication.BinlogSyncerConfig
+	localBinlogDir string
+	timezone       *time.Location
+
+	streamer         streamer.Streamer
+	streamerProducer StreamerProducer
+
+	// meetError means meeting error when get binlog event
+	// if binlogType is local and meetError is true, then need to create remote binlog stream
+	meetError bool
+
+	fromDB *UpStreamConn
+
+	uuidSuffix string
+}
+
+// NewStreamerController creates a new streamer controller
+func NewStreamerController(tctx tcontext.Context, syncCfg replication.BinlogSyncerConfig, fromDB *UpStreamConn, binlogType BinlogType, localBinlogDir string, timezone *time.Location, beginPos mysql.Position) (*StreamerController, error) {
+	streamerController := &StreamerController{
+		initBinlogType:    binlogType,
+		currentBinlogType: binlogType,
+		syncCfg:           syncCfg,
+		localBinlogDir:    localBinlogDir,
+		timezone:          timezone,
+		fromDB:            fromDB,
+	}
+
+	err := streamerController.ResetReplicationSyncer(tctx, beginPos)
+	if err != nil {
+		streamerController.Close(tctx)
+		return nil, err
+	}
+
+	return streamerController, nil
+}
+
+// ResetReplicationSyncer reset the replication
+func (c *StreamerController) ResetReplicationSyncer(tctx tcontext.Context, pos mysql.Position) (err error) {
+	c.Lock()
+	defer c.Unlock()
+
+	uuidSameWithUpstream := true
+
+	// close old streamerProducer
+	if c.streamerProducer != nil {
+		switch t := c.streamerProducer.(type) {
+		case *remoteBinlogReader:
+			c.closeBinlogSyncer(tctx, t.reader)
+		case *localBinlogReader:
+			// check the uuid before close
+			uuidSameWithUpstream, err = c.checkUUIDSameWithUpstream(pos, t.reader.GetUUIDs())
+			if err != nil {
+				return err
+			}
+			// TODO: close old local reader before creating a new one
+		default:
+			// some other producers such as mockStreamerProducer, should not re-create
+			c.streamer, err = c.streamerProducer.generateStreamer(pos)
+			return err
+		}
+	}
+
+	// binlog type is local: means relay already use the server id, so need change to a new server id
+	// binlog type is remote: dm-worker has more than one sub task, so need generate random server id
+	randomServerID, err := utils.GetRandomServerID(tctx.Context(), c.fromDB.BaseDB.DB)
+	if err != nil {
+		// should never happened unless the master has too many slave
+		return terror.Annotate(err, "fail to get random server id for streamer controller")
+	}
+	c.syncCfg.ServerID = randomServerID
+
+	// re-create new streamerProducer
+	if c.initBinlogType == LocalBinlog {
+		if c.meetError {
+			// meetError is true means meets error when get binlog event, in this case use remote binlog as default
+			if !uuidSameWithUpstream {
+				// if the binlog position's uuid is different from the upstream, can not switch to remote binlog
+				tctx.L().Info("may switch master in upstream, so can not switch local to remote")
```

Comment:
Is `Warn` better than `Info` here? the same with L175.

Reply:
update fff2db7